httpclient5/src/main/java/org/apache/hc/client5/http/auth/BearerToken.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+package org.apache.hc.client5.http.auth;
+
+import java.io.Serializable;
+import java.security.Principal;
+import java.util.Objects;
+
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+import org.apache.hc.core5.util.Args;
+
+/**
+ * Opaque token {@link Credentials} usually representing a set of claims, often encrypted
+ * or signed. The JWT (JSON Web Token) is among most widely used tokens used at the time
+ * of writing.
+ *
+ * @since 5.3
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public class BearerToken implements Credentials, Serializable {
+
+    private final Principal principal;
+    private final String token;
+
+    /**
+     * The constructor with the user principal and token arguments.
+     *
+     * @param userPrincipal the user principal
+     * @param token the token
+     *
+     * @see BasicUserPrincipal
+     * @see NTUserPrincipal
+     */
+    public BearerToken(final Principal userPrincipal, final String token) {
+        super();
+        this.principal = Args.notNull(userPrincipal, "User principal");
```

Comment:
@michael-o Why would not principal make sense here? Credentials are always associated with a particular security principal of some sort. Whether or not we can parse out the principal out of the token itself is a whole different story.

Reply:
This might be or not. The RFC does not indicate whether a subject is associated with. It solely talks about resource access.
Now let's look at this from client's perspective: I cannot obtain the principal since the token is supposed to be opaque to me and I have received it from my OIDC provider. Even with client credentials there is no guarantee that the final subject will correspond to the client_id? Should I supply a bogus principal?

Reply:
@michael-o No matter what perspective, HTTP requests are always executed on someone's behalf and tokens is not something you find lying on the ground. They are acquired for a specific principal. It is bogus only if you make it bogus.

Reply:
I understand that, but what do you expect people to supply if they cannot deliver that?

Reply:
@michael-o What do you mean? They do not know what they are doing? HTTP requests are no farts. They do not happen once of a sudden and for no reason. They are always executed by some agent for a specific means. There is always at least the application with an application id of some sort.

Reply:
Yes, but my understanding is that the token is quite abstract compared to Basic.

Reply:
@michael-o It is, but despite being abstract it still represents a principal. Would you rather like it to be null?

Reply:
From my point it can be `null` since we don't have it [here](https://github.com/apache/httpcomponents-client/blob/5.3.x/httpclient5/src/main/java/org/apache/hc/client5/http/auth/KerberosCredentials.java) as well. (here it could ge source with the `GSSName`, but that is a different discussion)